File: medium/210-Course-Schedule-II/TopologicalSort.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        Map<Integer, Set<Integer>> graph = new HashMap<>();
        for (int i = 0; i < numCourses; i++) {
            graph.put(i, new HashSet<Integer>());
        }
        
        int[] indegree = new int[numCourses];
        
        for (int[] preq : prerequisites) {
            graph.get(preq[1]).add(preq[0]);
            ++indegree[preq[0]];
        }
        
        return topSort(graph, indegree);
    }
    
    public int[] topSort(Map<Integer, Set<Integer>> graph, int[] indegree) {
        Stack<Integer> stack = new Stack<>();
        List<Integer> order = new ArrayList<>();
        
        for (int i = 0; i < indegree.length; i++) {
            if (indegree[i] == 0) stack.push(i);
        }
        
        while (!stack.isEmpty()) {
            int node = stack.pop();
            order.add(node);
            for (int adjacent: graph.get(node)) {
                if (--indegree[adjacent] == 0) stack.push(adjacent);
            }
        }
        
        return (order.size() == indegree.length) ? order.stream().mapToInt(i -> i).toArray() : new int[0];
    }
}

File: medium/210-Course-Schedule-II/TopologicalSort.kt
class Solution {
    fun findOrder(numCourses: Int, prerequisites: Array<IntArray>): IntArray {
        val graph = mutableMapOf<Int, MutableSet<Int>>()
        val indegree = IntArray(numCourses) {0}
        
        for (i in 0 until numCourses) {
            graph[i] = mutableSetOf<Int>()
        }
        
        for (pair in prerequisites) {
            graph[pair[1]]!!.add(pair[0])
            ++indegree[pair[0]]
        }
        
        return topSort(graph, indegree)
    }
    
    fun topSort(graph: Map<Int, Set<Int>>, indegree: IntArray): IntArray {
        val stack = Stack<Int>()
        indegree.forEachIndexed {i, v -> if (v == 0) stack.push(i)}
        
        var order = mutableListOf<Int>()
        while (!stack.isEmpty()) {
            val node = stack.pop()
            order.add(node)
            for (adjacent in graph[node]!!) {
                if (--indegree[adjacent] == 0) stack.push(adjacent);
            }
        }
        
        return if (order.size == indegree.size) order.toIntArray() else IntArray(0)
    }
}
